src/output.rs

```diff
@@ -321,56 +365,87 @@ where
                 loop {
-                    let (origin, info) = match
-                        self.snapshot.as_ref().payload().get(index)
-                    {
-                        Some(&(Payload::Origin(origin), ref info)) => {
-                            (origin, info)
-                        },
-                        Some(_) => continue,
+                    let (origin, info) = match iter.next_with_info() {
+                        Some((origin, info)) => (origin, info),
+                        None => {
+                            self.formatter.intermission(target)?;
+                            break
+                        }
+                    };
+                    if let Some(selection) = self.selection.as_ref() {
+                        if !selection.include_origin(origin) {
+                            continue
+                        }
+                    }
+                    if *first {
+                        *first = false;
+                    }
+                    else {
+                        self.formatter.delimiter(target)?;
+                    }
+                    self.formatter.origin(origin, info, target)?;
+                    return Ok(true)
+                }
+                (
+                    true,
+                    StreamState::Key {
+                        iter: self.snapshot.clone().arc_router_keys_iter(),
+                        first: true
+                    }
+                )
+            }
+            StreamState::Key { ref mut iter, ref mut first } => {
+                loop {
+                    let (key, info) = match iter.next_with_info() {
+                        Some((key, info)) => (key, info),
                         None => {
                             self.formatter.footer(
                                 self.metrics.as_ref(), target
                             )?;
-                            self.state = StreamState::Done;
                             break
                         }
                     };
-                    index += 1;
                     if let Some(selection) = self.selection.as_ref() {
-                        if !selection.as_ref().include_origin(origin) {
+                        if !selection.include_router_key(key) {
                             continue
                         }
                     }
-                    if !first {
+                    if *first {
+                        *first = false;
+                    }
+                    else {
                         self.formatter.delimiter(target)?;
                     }
-                    self.formatter.origin(origin, info, target)?;
-                    self.state = StreamState::Origin { index, first: false };
-                    break
+                    self.formatter.router_key(key, info, target)?;
+                    return Ok(true)
                 }
-                Ok(true)
+                (true, StreamState::Done)
```

Comment:
Seems like StreamState::Done has no functional use (anymore) and can be removed from the match if we simply return here:
```suggestion
               return Ok(false)
```

Reply:
It actually is still necessary. The stream essentially is an iterator but instead of returning an item, it writes it to the target. The return value of the method is whether there was a value. Because the stream writes the footer if it runs out of items, it needs to return `true` after that and can only return `false` if it is called once more. `StreamState::Done` signals that that has to happen next time.